posts/2022/09-15-testing-async.md

```diff
@@ -0,0 +1,82 @@
+==title==
+Testing Async Processes in Elixir
+
+==author==
+Alan Vardy
+
+==footer==
+
+==description==
+Improve the testing of async processes such as GenServers, Tasks, and GenStage pipelines.
+
+==tags==
+elixir,testing
+
+==body==
+
+Testing async processes in Elixir can be complex, especially around database calls and timing issues. There are a few principles that can make it easier to write dependable and fast tests.
+
+## Start processes in the test case
+
+While we normally want to start our processes in the `Application` module, in the test environment it is beneficial to start processes manually with `start_supervised!/2` in the test case setup. This allows us to have control over when a process starts and what state it starts with.
+
+You can make the starting of a process conditional by wrapping it in an environment check that is evaluated at compile time.
+
+```elixir
+# Application
+
+if Mix.env() === :test do
+  def processes, do: []
+else
+  def processes, do: [{SomeProcess, self()}]
+end
+```
+
+and then start the process in your test case
+
+```elixir
+# your test file
+
+setup do
+  start_supervised!({SomeProcess, self()})
+end
+```
+
+## Pass in the caller to fix Ecto sandbox issues
+
+Database calls are typically run within a transaction in the test environment so that any changes made during the test can be rolled back at the end. By default, one transaction is taken out per test case and it is linked to the PID of the test. This is called `manual` mode and is used when the `use DataCase` macro is set to `async: true`. The alternative mode is called `shared` and is used when the macro is set to `async: false` or just omitted entirely. In shared mode all processes running at the same time share the same transaction. Shared mode and synchronous tests go together because async tests need individual transactions to make sure that they are not sharing state. We want test transactions to be separate.
+
+When running an async process in manual mode, any database calls it makes will error because there was no transaction taken out for that PID (only for the main test process). It would be easy to set `async: false` on the `use DataCase` macro to resolve the issue, but the trouble with setting these tests to synchronous is that it dramatically slows down the runtime of the test suite, and we just need to come back and fix it later anyways.
+
+The solution is to use the `caller` option of Repo. If we pass in the PID of the main test process it will use that PID when finding the transaction rather than its own.
+
+```elixir
+MyApp.Repo.all(from(u in User), caller: test_pid)
+```
+
+When starting up the async process in `start_supervised!/2`, pass in the PID of the test process and use that for all repo calls. You can see an example of this in the last section.
```

Comment:
how do i get the test_pid?
Is it just:

setup do
  start_supervised!({SomeProcess, test_pid})
end

?